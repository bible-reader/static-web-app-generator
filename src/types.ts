import { BibleVersion, Chapter } from "@bible-reader/types";

export interface AppState {
  passages: Chapter[];
}

export interface BibleVersionWithHashes extends BibleVersion {
  hashes: BibleHashes;
}

export interface BibleVersionsMap {
  [bibleId: string]: BibleVersionWithHashes;
}

export interface BibleInputConfig {
  id: string;
  lang: string;
  name: string;
  input: string;
  pathInArchive?: string;
  type: string;
}

export type StateReducerFunc = (state: AppState) => AppState;

export type PassageAddFunc = () => void;

export type PassageChangeCreatorFunc = (
  passageIndex: number,
  versionId: string,
  book: string,
  chapter: number,
  verses: string[],
  loading: boolean
) => StateReducerFunc;

export type PassageChangeFunc = (
  index: number,
  versionId: string,
  book: string,
  chapter: number
) => void;

export type PassageChangeFuncCurried = (
  versionId: string,
  book: string,
  chapter: number
) => void;

export type PassageCloseCreatorFunc = (
  passageIndex: number
) => StateReducerFunc;

export type PassageCloseFunc = (index: number) => void;

export type PassageCloseFuncCurried = () => void;

type Hash = string;

interface BibleHashes {
  allHash?: Hash;
  v11nHash?: Hash;
  booksHashes?: {
    [bookId: string]: Hash;
  };
}

export interface BiblesHashes {
  [bibleId: string]: BibleHashes;
}
