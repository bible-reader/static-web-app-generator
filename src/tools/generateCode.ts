import { ChapterReference } from "@bible-reader/types";
import { BibleInputConfig, BiblesHashes } from "../types";

import { pad } from "../utils/utils";

function generateCode(
  publicPath: string,
  bibles: BibleInputConfig[],
  { versionId, book, chapter }: ChapterReference,
  biblesJsonHash: string,
  biblesHashes: BiblesHashes
) {
  const biblesJSON = require(`${publicPath}/bibles.${biblesJsonHash}.json`);
  let code = `/**
 * This file is generated using npm run generate and is not to be manualy modified.
 */
 /* tslint:disable */
import { ChapterContent } from "@bible-reader/types";
import { BibleVersionsMap } from "./types";
`;
  const descriptorHash = biblesHashes[versionId].descriptorHash;
  const descriptor = require(`${publicPath}/${versionId}/descriptor.${descriptorHash}.json`);

  bibles.forEach(({ id }: BibleInputConfig) => {
    biblesJSON[id].v11n = {}; // v11n data for current bible will be loaded right after first page load
  });
  code += `const bibles: BibleVersionsMap = ${JSON.stringify(biblesJSON)};\n`;

  const hash = descriptor.chapters[book][chapter - 1];
  const initialChapterJSON = require(`${publicPath}/${versionId}/${book}/ch${pad(
    chapter.toString(),
    3
  )}.${hash}.json`);

  code += `const initialChapter: ChapterContent`;
  code += `  = ${JSON.stringify(initialChapterJSON)};\n`;
  code += "export default {\n";
  code += "  bibles,\n";
  code += "  defaultPassage: {\n";
  code += `    versionId: "${versionId}",\n`;
  code += `    book: "${book}",\n`;
  code += `    chapter: ${chapter},\n`;
  code += `    verses: initialChapter.verses,\n`;
  code += `    loading: false\n`;
  code += "  }\n";
  code += "};\n";
  return code;
}

export default generateCode;
