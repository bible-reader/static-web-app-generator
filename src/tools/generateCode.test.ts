import { ChapterReference } from "@scripture-app/types";

import generateCode from "./generateCode";

const bibles = [
  {
    id: "cz_bkr",
    lang: "cz",
    name: "Bible KralickÃ¡",
    short: "BKR",
    input: "./src/tools/testData/czech_bkr_utf8_sample.txt",
    type: "unbound"
  },
  {
    id: "en_kjv",
    lang: "en",
    name: "King James Version",
    short: "KJV",
    input: "./src/tools/testData/KJV_test_sample.xmm",
    type: "opensong"
  }
];

const defaultChapter: ChapterReference = {
  versionId: "en_kjv",
  book: "gen",
  chapter: 1
};

const path = "../../src/tools/testData/bibles/";

describe("generateCode", () => {
  it("should generate correct code", () => {
    expect(generateCode(path, bibles, defaultChapter)).toMatchSnapshot();
  });
});
