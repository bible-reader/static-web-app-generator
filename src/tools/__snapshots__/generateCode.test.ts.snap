// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateCode should generate correct code 1`] = `
"/**
 * This file is generated using npm run generate and is not to be manualy modified.
 */
 /* tslint:disable */
import { ChapterContent } from \\"@bible-reader/types\\";
import { BibleVersionsMap } from \\"./types\\";
const bibles: BibleVersionsMap = {\\"kjv\\":{\\"id\\":\\"kjv\\",\\"lang\\":\\"en\\",\\"name\\":\\"King James Version\\",\\"hashes\\":{\\"allHash\\":\\"edb19a\\",\\"v11nHash\\":\\"c9bbd0\\",\\"booksHashes\\":{\\"gen\\":\\"65bd48\\",\\"exo\\":\\"e22dea\\",\\"lev\\":\\"bbdf85\\",\\"num\\":\\"bbdf85\\",\\"deu\\":\\"bbdf85\\"}},\\"v11n\\":{\\"gen\\":[31,25,24,26,32,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],\\"exo\\":[22,25,22,31,23,30,25,32,35,29,10,51,22,31,27,36,16,27,25,26,36,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],\\"lev\\":[1,1,1],\\"num\\":[1,1,1],\\"deu\\":[1,1,1]}},\\"bkr\\":{\\"id\\":\\"bkr\\",\\"lang\\":\\"cz\\",\\"name\\":\\"Bible Kralick√°\\",\\"hashes\\":{\\"allHash\\":\\"c678a9\\",\\"v11nHash\\":\\"d21033\\",\\"booksHashes\\":{\\"gen\\":\\"a53443\\",\\"exo\\":\\"a8888d\\"}},\\"v11n\\":{\\"gen\\":[31,25,24,26,32,22,24,22,29,32,32,20,18,24,21,16,27,33,38,18,34,24,20,67,34,35,46,22,35,43,55,32,20,31,29,43,36,30,23,23,57,38,34,34,28,34,31,22,33,26],\\"exo\\":[22,25,22,31,23,30,25,32,35,29,10,51,22,31,27,36,16,27,25,26,36,31,33,18,40,37,21,43,46,38,18,35,23,35,35,38,29,31,43,38]}}};
const initialChapter: ChapterContent  = {\\"verses\\":[\\"In the beginning God created the heaven and the earth.\\",\\"And the earth was without form, and void; and darkness was upon the face of the deep. And the Spirit of God moved upon the face of the waters.\\",\\"And God said, Let there be light: and there was light.\\",\\"And God saw the light, that it was good: and God divided the light from the darkness.\\",\\"And God called the light Day, and the darkness he called Night. And the evening and the morning were the first day.\\",\\"And God said, Let there be a firmament in the midst of the waters, and let it divide the waters from the waters.\\",\\"And God made the firmament, and divided the waters which were under the firmament from the waters which were above the firmament: and it was so.\\",\\"And God called the firmament Heaven. And the evening and the morning were the second day.\\",\\"And God said, Let the waters under the heaven be gathered together unto one place, and let the dry land appear: and it was so.\\",\\"And God called the dry land Earth; and the gathering together of the waters called he Seas: and God saw that it was good.\\",\\"And God said, Let the earth bring forth grass, the herb yielding seed, and the fruit tree yielding fruit after his kind, whose seed is in itself, upon the earth: and it was so.\\",\\"And the earth brought forth grass, and herb yielding seed after his kind, and the tree yielding fruit, whose seed was in itself, after his kind: and God saw that it was good.\\",\\"And the evening and the morning were the third day.\\",\\"And God said, Let there be lights in the firmament of the heaven to divide the day from the night; and let them be for signs, and for seasons, and for days, and years:\\",\\"And let them be for lights in the firmament of the heaven to give light upon the earth: and it was so.\\",\\"And God made two great lights; the greater light to rule the day, and the lesser light to rule the night: he made the stars also.\\",\\"And God set them in the firmament of the heaven to give light upon the earth,\\",\\"And to rule over the day and over the night, and to divide the light from the darkness: and God saw that it was good.\\",\\"And the evening and the morning were the fourth day.\\",\\"And God said, Let the waters bring forth abundantly the moving creature that hath life, and fowl that may fly above the earth in the open firmament of heaven.\\",\\"And God created great whales, and every living creature that moveth, which the waters brought forth abundantly, after their kind, and every winged fowl after his kind: and God saw that it was good.\\",\\"And God blessed them, saying, Be fruitful, and multiply, and fill the waters in the seas, and let fowl multiply in the earth.\\",\\"And the evening and the morning were the fifth day.\\",\\"And God said, Let the earth bring forth the living creature after his kind, cattle, and creeping thing, and beast of the earth after his kind: and it was so.\\",\\"And God made the beast of the earth after his kind, and cattle after their kind, and every thing that creepeth upon the earth after his kind: and God saw that it was good.\\",\\"And God said, Let us make man in our image, after our likeness: and let them have dominion over the fish of the sea, and over the fowl of the air, and over the cattle, and over all the earth, and over every creeping thing that creepeth upon the earth.\\",\\"So God created man in his own image, in the image of God created he him; male and female created he them.\\",\\"And God blessed them, and God said unto them, Be fruitful, and multiply, and replenish the earth, and subdue it: and have dominion over the fish of the sea, and over the fowl of the air, and over every living thing that moveth upon the earth.\\",\\"And God said, Behold, I have given you every herb bearing seed, which is upon the face of all the earth, and every tree, in the which is the fruit of a tree yielding seed; to you it shall be for meat.\\",\\"And to every beast of the earth, and to every fowl of the air, and to every thing that creepeth upon the earth, wherein there is life, I have given every green herb for meat: and it was so.\\",\\"And God saw every thing that he had made, and, behold, it was very good. And the evening and the morning were the sixth day.\\"]};
export default {
  bibles,
  defaultPassage: {
    versionId: \\"kjv\\",
    book: \\"gen\\",
    chapter: 1,
    verses: initialChapter.verses,
    loading: false
  }
};
"
`;
